File: Firmware/Core/Src/kinematics.c
/**
 * @file kinematics.c
 * @brief Implementación analítica del modelado cinemático directo e inverso.
 * @author Matías Exequiel Molina <ingenieria@uncuyo.edu.ar>
 * @date 2026
 *
 * @details Este módulo resuelve el mapeo cinemático para el manipulador paralelo de 3 GDL
 * (arquitectura EEZYbotARM MK2 modificada). Utiliza una formulación trigonométrica cerrada
 * en el plano proyectado r-z' generada a partir de los parámetros constructivos validados en CAD.
 *
 * @note Principios matemáticos implementados:
 *  - Eje 1 (Cintura): Rotación azimutal cartesiana en el plano X-Y.
 *  - Ejes 2 y 3 (Hombro y Muñeca): Desacople en el plano sagital r-z' mediante la Ley del Coseno.
 *  - Eliminación de articulaciones virtuales D-H: El paralelismo inducido por el mecanismo de
 *    cuatro barras se resuelve de forma directa manteniendo la orientación horizontal del efector.
 */

#include "kinematics.h"
#include <math.h>

#ifndef M_PI_F
#define M_PI_F (3.14159265358979323846f)
#endif

/**
 * @brief Evalúa si un vector de configuración articular respeta los topes mecánicos del robot.
 *
 * @param[in] q_deg Arreglo de 3 elementos con las coordenadas articulares [Q1, Q2, Q3] en grados [°].
 *
 * @return true  Las 3 articulaciones están estrictamente dentro de sus cotas operativas.
 * @return false Al menos una articulación excede su límite admisible.
 *
 * @note Límites configurados:
 *       - Q1: [-90.0°, +90.0°]
 *       - Q2: [+40.0°, +157.0°]
 *       - Q3: [-75.0°, +25.0°]
 */
bool Kinematics_CheckJointLimits(const float q_deg[3]) {
    if ((q_deg[0] < Q1_MIN_DEG) || (q_deg[0] > Q1_MAX_DEG)) {
        return false;
    }
    if ((q_deg[1] < Q2_MIN_DEG) || (q_deg[1] > Q2_MAX_DEG)) {
        return false;
    }
    if ((q_deg[2] < Q3_MIN_DEG) || (q_deg[2] > Q3_MAX_DEG)) {
        return false;
    }
    return true;
}

/**
 * @brief Resuelve la cinemática directa analítica (Forward Kinematics).
 *
 * @details Transforma las coordenadas generalizadas articulares [Q1, Q2, Q3] a la posición
 * tridimensional [X, Y, Z] del centro del efector final (solenoide) referida a la base.
 *
 * Ecuaciones:
 * \f[
 *   R = L_2 \cos(Q_2) + L_3 \cos(Q_3) + O_r
 * \f]
 * \f[
 *   X = R \cos(Q_1), \quad Y = R \sin(Q_1)
 * \f]
 * \f[
 *   Z = L_1 + L_2 \sin(Q_2) + L_3 \sin(Q_3) - O_z
 * \f]
 *
 * @param[in]  q_deg   Ángulos absolutos de las articulaciones [Q1, Q2, Q3] en grados sexagesimales [°].
 * @param[out] pos_mm  Vector cartesiano resultante [X, Y, Z] en milímetros [mm].
 *
 * @return true  Cálculo completado correctamente.
 * @note Esta función es puramente determinista y reentrante (thread-safe).
 */
bool Kinematics_FK(const float q_deg[3], float pos_mm[3]) {
    /* Conversión de unidades de entrada sexagesimales a radianes */
    const float q1_rad = q_deg[0] * DEG_TO_RAD_F;
    const float q2_rad = q_deg[1] * DEG_TO_RAD_F;
    const float q3_rad = q_deg[2] * DEG_TO_RAD_F;

    /* Proyección del radio planar total, integrando el offset constructivo del cabezal Or */
    const float r_planar = (KIN_L2_MM * cosf(q2_rad)) + (KIN_L3_MM * cosf(q3_rad)) + KIN_O_R_MM;

    /* Proyección cartesiana tridimensional referida al centro del eje de cintura */
    pos_mm[0] = r_planar * cosf(q1_rad);
    pos_mm[1] = r_planar * sinf(q1_rad);
    pos_mm[2] = KIN_L1_MM + (KIN_L2_MM * sinf(q2_rad)) + (KIN_L3_MM * sinf(q3_rad)) - KIN_O_Z_MM;

    return true;
}

/**
 * @brief Resuelve la cinemática inversa analítica geométrica (Inverse Kinematics).
 *
 * @details Determina la configuración articular requerida para alcanzar una posición objetivo [X, Y, Z].
 * Implementa la solución de rama geométrica "codo arriba" (elbow-up) de manera analítica y única,
 * evitando iteraciones y singularidades cinemáticas.
 *
 * Pasos algorítmicos:
 *  1. Cálculo de Q1 mediante proyección azimutal atan2(Y, X).
 *  2. Desacople de los offsets constructivos del efector (Or, Oz) para hallar el nodo de muñeca (r_wrist, z').
 *  3. Verificación de la distancia euclídea D contra el espacio de trabajo admisible (|L2 - L3| <= D <= L2 + L3).
 *  4. Cálculo de los ángulos internos beta y gamma mediante el Teorema del Coseno.
 *  5. Composición de Q2 = alpha + beta y Q3 = alpha - gamma.
 *  6. Validación final de los ángulos resultantes contra límites físicos de carrera.
 *
 * @param[in]  pos_mm  Coordenadas cartesianas deseadas [X, Y, Z] en milímetros [mm].
 * @param[out] q_deg   Arreglo de 3 elementos donde se retornan los ángulos [Q1, Q2, Q3] en grados [°].
 *
 * @return true  Pose cartesiana admisible y solución articular físicamente ejecutable.
 * @return false Punto fuera del volumen de trabajo o que viola los límites angulares.
 *
 * @warning Si el punto solicitado cae en la singularidad cilíndrica central (R_obj <= Or + 5 mm),
 *          la función aborta inmediatamente retornando false.
 */
bool Kinematics_IK(const float pos_mm[3], float q_deg[3]) {
    const float x = pos_mm[0];
    const float y = pos_mm[1];
    const float z = pos_mm[2];

    /* -------------------------------------------------------------------------
     * PASO 1: Rotación de base azimutal (Articulación Q1)
     * ------------------------------------------------------------------------- */
    const float q1_rad = atan2f(y, x);
    const float r_cyl = sqrtf((x * x) + (y * y));

    /* Aislamiento preventivo: evita singularidad de indeterminación en el eje vertical */
    if (r_cyl <= (KIN_O_R_MM + 5.0f)) {
        return false;
    }

    /* -------------------------------------------------------------------------
     * PASO 2: Traslación al centro del nodo de articulación de la muñeca
     * ------------------------------------------------------------------------- */
    const float r_wrist = r_cyl - KIN_O_R_MM;
    const float z_prime = z + KIN_O_Z_MM - KIN_L1_MM;

    /* -------------------------------------------------------------------------
     * PASO 3: Distancia euclídea D entre origen del hombro y nodo de muñeca
     * ------------------------------------------------------------------------- */
    const float d_dist = sqrtf((r_wrist * r_wrist) + (z_prime * z_prime));

    /* Validación del volumen de trabajo esférico permitido */
    if ((d_dist > (KIN_L2_MM + KIN_L3_MM)) || (d_dist < fabsf(KIN_L2_MM - KIN_L3_MM))) {
        return false;
    }

    /* -------------------------------------------------------------------------
     * PASO 4: Resolución planar 2D mediante Teorema del Coseno
     * ------------------------------------------------------------------------- */
    /* Ángulo interno beta (entre vector D y eslabón principal L2) */
    float cos_beta = ((d_dist * d_dist) + (KIN_L2_MM * KIN_L2_MM) - (KIN_L3_MM * KIN_L3_MM)) /
                     (2.0f * d_dist * KIN_L2_MM);
    if (cos_beta > 1.0f)  cos_beta = 1.0f;
    if (cos_beta < -1.0f) cos_beta = -1.0f;
    const float beta = acosf(cos_beta);

    /* Ángulo de elevación alpha del vector D respecto al horizonte */
    const float alpha = atan2f(z_prime, r_wrist);

    /* Ángulo interno gamma (entre vector D y antebrazo L3) */
    float cos_gamma = ((d_dist * d_dist) + (KIN_L3_MM * KIN_L3_MM) - (KIN_L2_MM * KIN_L2_MM)) /
                      (2.0f * d_dist * KIN_L3_MM);
    if (cos_gamma > 1.0f)  cos_gamma = 1.0f;
    if (cos_gamma < -1.0f) cos_gamma = -1.0f;
    const float gamma = acosf(cos_gamma);

    /* -------------------------------------------------------------------------
     * PASO 5: Composición angular absoluta de la cadena cinemática ("Codo Arriba")
     * ------------------------------------------------------------------------- */
    const float q2_rad = alpha + beta;
    const float q3_rad = alpha - gamma;

    /* Conversión de radianes a grados sexagesimales */
    q_deg[0] = q1_rad * RAD_TO_DEG;
    q_deg[1] = q2_rad * RAD_TO_DEG;
    q_deg[2] = q3_rad * RAD_TO_DEG;

    /* -------------------------------------------------------------------------
     * PASO 6: Validación de topes mecánicos físicos
     * ------------------------------------------------------------------------- */
    return Kinematics_CheckJointLimits(q_deg);
}

File: ros2_ws/src/robot_control/robot_control/gui/kinematics.py
"""Modelado cinemático analítico para el manipulador paralelo de 3 GDL.

Implementa la cinemática directa (FK) e inversa (IK) geométrica en el
plano sagital proyectado r-z', eliminando la necesidad de eslabones y
articulaciones virtuales de la convención Denavit-Hartenberg.
"""

from typing import List, Tuple
import numpy as np

# =========================================================================
# PARÁMETROS GEOMÉTRICOS Y OFFSETS CONSTRUCTIVOS [m]
# =========================================================================
L1_M: float = 0.07700   # Altura axial de la base al centro del hombro [m]
L2_M: float = 0.13500   # Eslabón principal de elevación (hombro a codo) [m]
L3_M: float = 0.14700   # Eslabón de antebrazo (codo a muñeca) [m]
O_R_M: float = 0.02549  # Desfase radial constructivo del efector final [m]
O_Z_M: float = 0.04087  # Desfase axial vertical del efector final [m]

# Límites angulares de seguridad física articular [°]
Q1_MIN_DEG: float = -90.0
Q1_MAX_DEG: float = 90.0
Q2_MIN_DEG: float = 40.0   # Mantiene coherencia estricta con el firmware
Q2_MAX_DEG: float = 157.0
Q3_MIN_DEG: float = -75.0
Q3_MAX_DEG: float = 25.0


def forward_kinematics(
    q_deg: List[float],
) -> Tuple[float, float, float, np.ndarray]:
    """Resuelve la cinemática directa analítica (FK).

    Calcula la posición cartesiana [X, Y, Z] del efector final referida al
    origen de la base a partir del vector de ángulos articulares.

    Args:
        q_deg (List[float]): Vector articular [Q1, Q2, Q3] en grados [°].

    Returns:
        Tuple[float, float, float, np.ndarray]:
            - x_mm (float): Coordenada longitudinal X [mm].
            - y_mm (float): Coordenada transversal Y [mm].
            - z_mm (float): Coordenada vertical Z [mm].
            - T_total (np.ndarray): Matriz de transformación homogénea 4x4 [m].
    """
    q_rad = np.radians(np.round(q_deg, 4))
    q1, q2 = float(q_rad[0]), float(q_rad[1])
    q3 = float(q_rad[2]) if len(q_rad) > 2 else 0.0

    # Radio proyectado planar integrando desfase del cabezal Or
    radius_m = L2_M * np.cos(q2) + L3_M * np.cos(q3) + O_R_M

    # Posición tridimensional referida a la base en metros
    x_m = radius_m * np.cos(q1)
    y_m = radius_m * np.sin(q1)
    z_m = L1_M + L2_M * np.sin(q2) + L3_M * np.sin(q3) - O_Z_M

    # Conversión métrica a milímetros
    x_mm = float(x_m * 1000.0)
    y_mm = float(y_m * 1000.0)
    z_mm = float(z_m * 1000.0)

    if abs(y_mm) < 1e-4:
        y_mm = 0.0

    # Construcción de la matriz homogénea cartesiana simplificada
    ct1, st1 = np.cos(q1), np.sin(q1)
    t_total = np.array(
        [
            [ct1, -st1, 0.0, x_m],
            [st1, ct1, 0.0, y_m],
            [0.0, 0.0, 1.0, z_m],
            [0.0, 0.0, 0.0, 1.0],
        ],
        dtype=np.float64,
    )

    return x_mm, y_mm, z_mm, t_total


def inverse_kinematics(
    x_mm: float, y_mm: float, z_mm: float, raw_float: bool = True
) -> Tuple[List[float], bool]:
    """Resuelve la cinemática inversa analítica geométrica (IK).

    Determina la solución articular para alcanzar el objetivo cartesiano,
    forzando la configuración geométrica de codo arriba ("elbow-up").

    Args:
        x_mm (float): Coordenada objetivo X en milímetros [mm].
        y_mm (float): Coordenada objetivo Y en milímetros [mm].
        z_mm (float): Coordenada objetivo Z en milímetros [mm].
        raw_float (bool, optional): Si es True retorna flotantes continuos;
            si es False redondea a dos decimales. Por defecto es True.

    Returns:
        Tuple[List[float], bool]:
            - q_deg (List[float]): Vector articular resuelto [Q1, Q2, Q3] en grados [°].
            - reachable (bool): True si la solución es alcanzable y físicamente válida;
                                False si cae fuera del workspace o viola límites.
    """
    x = x_mm / 1000.0
    y = y_mm / 1000.0
    z = z_mm / 1000.0

    # 1. Rotación azimutal de la base (Articulación Q1)
    q1_rad = np.arctan2(y, x)
    r_obj = np.hypot(x, y)

    # Protección contra indeterminación en eje central
    if r_obj <= (O_R_M + 0.005):
        return [0.0, 0.0, 0.0], False

    # 2. Desacople geométrico hacia el nodo de articulación de muñeca
    r_wrist = r_obj - O_R_M
    z_prime = z + O_Z_M - L1_M

    # 3. Distancia euclídea planar entre hombro y muñeca
    d_dist = np.hypot(r_wrist, z_prime)

    # Comprobación de alcanzabilidad en volumen esférico
    if d_dist > (L2_M + L3_M) or d_dist < abs(L2_M - L3_M):
        return [0.0, 0.0, 0.0], False

    # 4. Solución planar 2D mediante Teorema del Coseno
    cos_beta = np.clip(
        (d_dist**2 + L2_M**2 - L3_M**2) / (2.0 * d_dist * L2_M), -1.0, 1.0
    )
    beta = np.arccos(cos_beta)

    alpha = np.arctan2(z_prime, r_wrist)

    cos_gamma = np.clip(
        (d_dist**2 + L3_M**2 - L2_M**2) / (2.0 * d_dist * L3_M), -1.0, 1.0
    )
    gamma = np.arccos(cos_gamma)

    # 5. Ángulos absolutos en configuración codo arriba
    q2_rad = alpha + beta
    q3_rad = alpha - gamma

    q1_deg = float(np.degrees(q1_rad))
    q2_deg = float(np.degrees(q2_rad))
    q3_deg = float(np.degrees(q3_rad))

    # 6. Validación contra límites de carrera articulares
    if not (Q1_MIN_DEG <= q1_deg <= Q1_MAX_DEG):
        return [0.0, 0.0, 0.0], False
    if not (Q2_MIN_DEG <= q2_deg <= Q2_MAX_DEG):
        return [0.0, 0.0, 0.0], False
    if not (Q3_MIN_DEG <= q3_deg <= Q3_MAX_DEG):
        return [0.0, 0.0, 0.0], False

    if raw_float:
        return [q1_deg, q2_deg, q3_deg], True
    return [round(q1_deg, 2), round(q2_deg, 2), round(q3_deg, 2)], True